### PR TITLE
Adding `Resources` and `OpTypeWithParams` to labs

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
As part of adding the experimental changes to core PennyLane, this PR adds some container classes
for storing a light weight representation of the operator along with modifications to Resources class.

**Description of the Change:**
- Added `OpTypeWithParams`. 
- Modified `Resources` from core PennyLane.

### CODE REVIEW

File: pennylane/labs/resource_estimation/resource_contianer.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Base classes for the resource objects and resource estimation."""
+from collections import defaultdict
+from dataclasses import dataclass, field
+
+
+class OpTypeWithParams:
+
+    def __init__(self, op_type: str, params_tuple: tuple) -> None:
```
Comment (willjmax) [inline pennylane/labs/resource_estimation/resource_contianer.py:21]: I think the types could be a bit more descriptive. Instead of `op_type: str` we should pass in a callable that produces a PennyLane Operator from `params_tuple`. I'm thinking something like this `x = OpTypeWithParams(qml.QFT, (('wires', [0, 1]))` and the corresponding type annotation should look like `op_type: Callable[[tuple], Operator]`. With this change we can still access the string through the `__name__ method, but now we can do error handling by checking `issubclass(op_type, Operator)` to ensure that the instantiated object corresponds to a PennyLane Operator.

Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_contianer.py:21]: Ahh no, we don't want to do this. Imagine that we have an operation that acts on lots of wires, we would rather store `num_wires` then explicitly a list of all of the wire ids that are required to instantiate the operator. 

This is the difference between the parameters required for full decomposition vs parameters required for resource level decompositions.

Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_contianer.py:21]: We could potentially pass around references to the class itself. I don't know how well that plays with things like hashing. Since its easier to hash a tuple of strings and ints, then it is to hash classes in memory

Comment (willjmax) [reply pennylane/labs/resource_estimation/resource_contianer.py]: I think we should pass in the class itself as input, that way we can check to make sure it is an Operator. Then we can hash on the string `op_type.__name__`.

File: pennylane/labs/resource_estimation/resource_contianer.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        ...             )
+        >>> print(r)
+        wires: 2
+        gates: 2
+        gate_types:
+        {'Hadamard(num_wires=1)': 1, 'CNOT(num_wires=2)': 1}
+    """
+    num_wires: int = 0
+    num_gates: int = 0
+    gate_types: defaultdict = field(default_factory=lambda: defaultdict(int))
```
Comment (AntonNI8) [inline pennylane/labs/resource_estimation/resource_contianer.py]: Regarding wires, is there a reason we're still tracking the count of wires `num_wires` as opposed to jumping straight into tracking the wire indices - as a list or array or set, ie. `wires: List[int] = []`, `wires: set = set()` ?
Regarding `num_gates`, if this simply refers to the total # of gates in the resource instance (ie. the sum of all counts in the `gate_types` dictionary), it seems a bit redundant; we could consider removing it.
It can be useful to know the total # of gates, of course, but this could (alternatively) be achieved by providing a class method that simply performs the addition upon request, rather than having to track/update internally as you go.

Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_contianer.py]: *We discussed this offline:

The conclusion was to use number of wires for now, and if we decide to stick with the original wires proposal to revert the change in a follow up PR.

File: pennylane/labs/resource_estimation/resource_contianer.py (hunk 1 above)
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_contianer.py]: Added 👍🏼

**Comment by Jaybsoni** [comment]
[sc-76751]

File: pennylane/labs/tests/resource_estimation/test_resource_container.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+    _combine_dict,
+    _scale_dict,
+    add_in_parallel,
+    add_in_series,
+    mul_in_parallel,
+    mul_in_series,
+)
+
+
+class TestCompressedResourceOp:
```
Comment (willjmax) [inline pennylane/labs/tests/resource_estimation/test_resource_container.py:36]: These tests will need to be rewritten once our other PRs are merged. We decided to build `CompressedResourceOp` from `ResourceOperator` instead of PennyLane operators, and the tests will need to be updated. Another issue is that we haven't added a resource operator for QSVT, so we should rewrite these tests to only use operators we have access to.

I'm fine with the tests as they are for this PR. Once the `resource_constructor` branch is merged these tests will fail since there will be a check to make sure a resource operator is passed into the init of the compressed resource op. Maybe we can fix these tests within that PR?

File: pennylane/labs/resource_estimation/resource_container.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+    if in_place:
+        first.num_gates = new_gates
+        return first
+
+    return Resources(first.num_wires, new_gates, new_gate_types)
+
+
+def mul_in_parallel(first: Resources, scaler: int, in_place=False) -> Resources:
+    r"""Multiply two resources assuming the circuits are executed in parallel.
```
Comment (willjmax) [inline pennylane/labs/resource_estimation/resource_container.py:234]: Maybe reword this since we aren't multiplying two resources, we are multiplying a resource by a scalar.

File: pennylane/labs/resource_estimation/resource_container.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+            )
+        >>> print(op_tp)
+        QSVT(num_wires=5, num_angles=100)
+    """
+
+    def __init__(self, op_type: type, params: dict) -> None:
+        r"""Instantiate the light weight class corressponding to the operator type and parameters.
+
+        Args:
+            op_type (Type): the PennyLane type of the operation
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py:53]: Removed type hint and modified description

File: pennylane/labs/resource_estimation/resource_container.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+import copy
+from collections import defaultdict
+from dataclasses import dataclass, field
+
+
+class CompressedResourceOp:
+    r"""Instantiate the light weight class corressponding to the operator type and parameters.
+
+    Args:
+        op_type (Type): the PennyLane type of the operation
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py:24]: Modified description

File: pennylane/labs/resource_estimation/resource_container.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+        >>> op_tp = CompressedResourceOp(
+                qml.QSVT,
+                {
+                    "num_wires": 5,
+                    "num_angles": 100,
+                },
+            )
+        >>> print(op_tp)
+        QSVT(num_wires=5, num_angles=100)
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py:46]: Modified example. We can add a more detailed example once we merge the other PRs in

File: pennylane/labs/resource_estimation/resource_container.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+    Args:
+        first (Resources): first resource object to combine
+        other (Resources): other resource object to combine with
+        in_place (bool): determines if the first Resources are modified in place (default False)
+
+    Returns:
+        Resources: combined resources
+    """
+    new_wires = max(first.num_wires, other.num_wires)
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py:177]: This is a good question, we had initially suggested also tracking the wire labels explicitly. I ultimately decide to leave it out of scope so we can prioritize performance first. I think we can comeback and add the wire labels and modify these functions to be more accurate. For now this is a standard assumption for adding in parallel and in series from just the total number of wires.

File: pennylane/labs/resource_estimation/resource_container.py (hunk 4 above)
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py:234]: Added

File: pennylane/labs/tests/resource_estimation/test_resource_container.py (hunk 3 above)
Comment (Jaybsoni) [reply pennylane/labs/tests/resource_estimation/test_resource_container.py:36]: I think this is a good idea. Lets merge the constructor PR into this one, then I can modify the test and merge upwards 👍🏼

File: pennylane/labs/resource_estimation/resource_container.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+    __rmul__ = __mul__  # same implementation
+
+    def __iadd__(self, other: "Resources") -> "Resources":
+        """Add two resources objects in series"""
+        return add_in_series(self, other, in_place=True)
+
+    def __imull__(self, scalar: int) -> "Resources":
+        """Scale a resources object in series"""
+        return mul_in_series(self, scalar, in_place=True)
```
Comment (AntonNI8) [inline pennylane/labs/resource_estimation/resource_container.py:124]: Adding/multiplying in series makes sense as the default - we will have to ensure documentation makes it clear that this is what the dunder methods are doing.

**Review by AntonNI8 — APPROVED** [review]

File: pennylane/labs/__init__.py
Comment (soranjh) [inline pennylane/labs/__init__.py:1]: We should make labs importable by updating `pennylane/__init__.py`.

**Review by soranjh — COMMENTED** [review]
Thanks @Jaybsoni, left some suggestions. Couldn't run the examples though because of import issues.

File: pennylane/labs/__init__.py
Comment (Jaybsoni) [reply pennylane/labs/__init__.py:1]: No! There should be no Labs dependancies in core PL otherwise they could lead to experimental features in labs breaking core features.

File: pennylane/labs/resource_estimation/resource_container.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+from pennylane.labs.resource_estimation import ResourceOperator
+
+
+class CompressedResourceOp:
+    r"""Instantiate the light weight class corressponding to the operator type and parameters.
+
+    Args:
+        op_type: the class object for of an operation which inherits from '~.ResourceOperator'
+        params (dict): a dictionary containing the minimal pairs of parameter names and values
+                    required to compute the resources for the given operator!
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py:28]: Yes

File: pennylane/labs/resource_estimation/resource_container.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+
+    .. details::
+
+        This representation is the minimal amount of information required to estimate resources for the operator.
+
+        **Example**
+
+        >>> op_tp = CompressedResourceOp(ResourceHadamard, {"num_wires":1})
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py:36]: Yes, it should work fine. There are no checks to validate the inputs. I think its fine to keep it like this because we don't want users to construct these CompressedResourceOp objects anyways. Its more for us to have an efficient internal representation for the object.

File: pennylane/labs/resource_estimation/resource_container.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+        wires: 2
+        gates: 2
+        gate_types:
+        {'Hadamard': 1, 'CNOT': 1}
+    """
+
+    num_wires: int = 0
+    num_gates: int = 0
+    gate_types: defaultdict = field(default_factory=lambda: defaultdict(int))
+
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py:112]: We have decided to initially exclude them as they don't add any useful information but add some overhead to maintain

File: pennylane/labs/resource_estimation/resource_container.py [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+        )
+        items += "\ngate_types:\n{" + gate_type_str + "}"
+        return items
+
+    def _ipython_display_(self):
+        """Displays __str__ in ipython instead of __repr__"""
+        print(str(self))
+
+
+def add_in_series(first: Resources, other: Resources, in_place=False) -> Resources:
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py:151]: Yes!

File: pennylane/labs/resource_estimation/resource_operator.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+                @classmethod
+                def resource_rep(cls, num_wires) -> CompressedResourceOp:
+                    params = {"num_wires": num_wires}
+                    return CompressedResourceOp(cls, params)
+    """
+
+    @staticmethod
+    @abstractmethod
+    def _resource_decomp(*args, **kwargs) -> Dict[CompressedResourceOp, int]:
+        """Returns the Resource object. This method is only to be used inside
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_operator.py:69]: Yes, this is more like a resource specific decomposition, not actually returning the resources.

File: pennylane/labs/resource_estimation/resource_operator.py
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_operator.py:1]: Good point, we can move them into the same files.

File: pennylane/labs/resource_estimation/resource_container.py [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+
+        **Example**
+
+        >>> op_tp = CompressedResourceOp(ResourceHadamard, {"num_wires":1})
+        >>> print(op_tp)
+        Hadamard(num_wires=1)
+    """
+
+    def __init__(self, op_type, params: dict) -> None:
+        r"""Instantiate the light weight class corressponding to the operator type and parameters.
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py]: Keeping it for now, I think it's harmless to have.

File: pennylane/labs/resource_estimation/resource_container.py [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+from dataclasses import dataclass, field
+
+from pennylane.labs.resource_estimation import ResourceOperator
+
+
+class CompressedResourceOp:
+    r"""Instantiate the light weight class corressponding to the operator type and parameters.
+
+    Args:
+        op_type: the class object for of an operation which inherits from '~.ResourceOperator'
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py]: Added 👍🏼

File: pennylane/labs/tests/resource_estimation/test_resource_container.py [hunk 17]
```diff
@@ -0,0 +1,315 @@
...
+            num_wires=2,
+            num_gates=6,
+            gate_types=defaultdict(int, {"RZ": 2, "CNOT": 1, "RY": 2, "Hadamard": 1}),
+        )
+
+        resultant_obj = add_in_series(resource_obj, other_obj, in_place=in_place)
+        assert resultant_obj == expected_res_obj
+
+        if in_place:
+            assert resultant_obj is resource_obj
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/labs/tests/resource_estimation/test_resource_container.py:167]: Yes, the first assert checks if the result was modified to be what we expect. The second asset using the `is` is checking if the python objects are the same (i.e we modified the original in memory as opposed to creating a new copy)

File: pennylane/labs/.pylintrc [hunk 18]
```diff
@@ -30,7 +30,7 @@ ignored-classes=numpy,scipy,autograd,toml,appdir,autograd.numpy,autograd.numpy.l
 # it should appear only once).
 # Cyclical import checks are disabled for now as they are frequently used in
 # the code base, but this can be removed in the future once cycles are resolved.
-disable=line-too-long,invalid-name,too-many-lines,redefined-builtin,too-many-locals,duplicate-code,cyclic-import,import-error,bad-option-value
+disable=line-too-long,invalid-name,too-many-lines,redefined-builtin,too-many-locals,duplicate-code,cyclic-import,import-error,bad-option-value,too-few-public-methods,
```
Comment (Jaybsoni) [inline pennylane/labs/.pylintrc:33]:
```suggestion
disable=line-too-long,invalid-name,too-many-lines,redefined-builtin,too-many-locals,duplicate-code,cyclic-import,import-error,bad-option-value,too-few-public-methods
```

File: pennylane/labs/resource_estimation/resource_operator.py [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Callable, Dict
+
+if TYPE_CHECKING:
+    from pennylane.labs.resource_estimation import CompressedResourceOp
+
+
+class ResourceOperator(ABC):
+    r"""This is an abstract class that defines the methods a PennyLane Operator
+    must implement in order to be used for resource estimation.
```
Comment (soranjh) [inline pennylane/labs/resource_estimation/resource_operator.py:26]:
```suggestion
    r"""Abstract class that defines the methods a PennyLane Operator
    must implement in order to be used for resource estimation.
```

**Review by AntonNI8 — APPROVED** [review]
No outstanding concerns.

File: pennylane/labs/resource_estimation/resource_operator.py [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+                    return gate_types
+
+                def resource_params(self, num_wires) -> dict:
+                    return {"num_wires": num_wires}
+
+                @classmethod
+                def resource_rep(cls, num_wires) -> CompressedResourceOp:
+                    params = {"num_wires": num_wires}
+                    return CompressedResourceOp(cls, params)
+    """
```
Comment (Jaybsoni) [inline pennylane/labs/resource_estimation/resource_operator.py:64]:
```suggestion
        extend ``qml.QFT`` for resource estimation.

        .. code-block:: python

            import pennylane as qml
            from pennylane.labs.resource_estimation import CompressedResourceOp, ResourceOperator

            class ResourceQFT(qml.QFT, ResourceOperator):

                @staticmethod
                def _resource_decomp(num_wires) -> Dict[CompressedResourceOp, int]:
                    gate_types = {}

                    hadamard = ResourceHadamard.resource_rep()
                    swap = ResourceSWAP.resource_rep()
                    ctrl_phase_shift = ResourceControlledPhaseShift.resource_rep()

                    gate_types[hadamard] = num_wires
                    gate_types[swap] = num_wires // 2
                    gate_types[ctrl_phase_shift] = num_wires*(num_wires - 1) // 2

                    return gate_types

                def resource_params(self, num_wires) -> dict:
                    return {"num_wires": num_wires}

                @classmethod
                def resource_rep(cls, num_wires) -> CompressedResourceOp:
                    params = {"num_wires": num_wires}
                    return CompressedResourceOp(cls, params)
    
        Which can be instantiated as a normal operation, but now contains the resources: 
    
        .. code:: bash
        
            >>> op = ResourceQFT(range(3))
            >>> op.resources(**op.resource_params())
            {Hadamard(): 3, SWAP(): 1, ControlledPhaseShift(): 3}
        
    """
```

File: pennylane/labs/resource_estimation/resource_container.py [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+                        required to compute the resources for the given operator
+
+        .. details::
+
+            This representation is the minimal amount of information required to estimate resources for the operator.
+
+            **Example**
+
+            >>> op_tp = CompressedResourceOp(ResourceHadamard, {"num_wires":1})
+            >>> print(op_tp)
```
Comment (soranjh) [inline pennylane/labs/resource_estimation/resource_container.py:56]: I got some errors with this.

File: pennylane/labs/resource_estimation/resource_operator.py [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+                def resource_rep(cls, num_wires) -> CompressedResourceOp:
+                    params = {"num_wires": num_wires}
+                    return CompressedResourceOp(cls, params)
+
+        Which can be instantiated as a normal operation, but now contains the resources:
+
+        .. code-block:: bash
+
+            >>> op = ResourceQFT(range(3))
+            >>> op.resources(**op.resource_params())
```
Comment (soranjh) [inline pennylane/labs/resource_estimation/resource_operator.py:70]: Couldn't run this too.

**Review by soranjh — APPROVED** [review]

File: pennylane/labs/resource_estimation/resource_container.py (hunk 21 above)
Comment (Jaybsoni) [reply pennylane/labs/resource_estimation/resource_container.py:56]: image
